added </option> as an optional ommited end tag

pollei opened the pull request:
as per http://www.w3.org/TR/html5/syntax.html#optional-tags
    modified:   src/HTMLParser.js
    modified:   test/test-slowparse.js

pollei committed:
added </option> as an optional ommited end tag
  as per http://www.w3.org/TR/html5/syntax.html#optional-tags
	modified:   src/HTMLParser.js
	modified:   test/test-slowparse.js

Pomax commented on src/HTMLParser.js:
```diff
@@ -126,7 +126,17 @@ module.exports = (function(){
+      "rtc": ["rb","rtc","rp"],
+      "rp": ["rb","rt","rtc","rp"],
+      "optgroup": ["optgroup"],
+      "option": ["option"],
```
or may also be followed by an "optgroup"

Pomax commented on test/test-slowparse.js:
```diff
@@ -597,6 +597,12 @@ module.exports = function(Slowparse, window, document, validators) {
+    var result = parse(html);
+    ok(!result.error, "no error on omitted </option>");
+  });
+
```
given the large number of entries in addition to `option` you added, it's important to add test cases for those, too.

Pomax commented:
super-thanks for adding those! A few notes on the PR though, mostly the fact that you added more than just `option` means we'll need tests for the other optional tag elements, too.

pollei commented:
OK I'll add more test cases, and I'll work to make them pass. Right now they will definitely fail if I'm reading the code correctly; I added them to high-light incompleteness. So I'll make tests for "tr", "optgroup" and the ruby stuff, etc.

I have no real clue about ruby. https://developer.mozilla.org/en-US/docs/Web/HTML/Element/ruby however I can find some examples and adapt them as tests. I'll cross-check with https://validator.w3.org/ .

https://bugzilla.mozilla.org/show_bug.cgi?id=946393 "Allow optional omittable close tags according to the W3C HTML5 standards"
I'll also add tests for "dd" and "dt" to help close this bug on thimble. It's an old one from 2013-12-04 .
Then thimble would have to update their package.json as they have "slowparse": "1.1.3" . and you are now 1.1.4 . likely you should go to version 1.1.5 soon.

50% chance I'll have another patch ready to review by end of tomorrow, 95% chance by Monday.

Pomax commented:
Thimble just relaunched a completely overhaul on thimble.mozilla.org, using a codebase that isn't making use of Slowparse at the moment, but adding it back in does seem on their roadmap from what I can tell.

humphd commented:
Indeed, I expect Slowparse to rejoin Thimble soon.  We just lack elegant UI, but patches like this make the backend something to be desired.

pollei committed:
Added more to omittableCloseTagHtmlElements
fixed dd and dt as per https://bugzilla.mozilla.org/show_bug.cgi?id=946393
got some ruby working
added more tests of omitable tags; 6 of which currently fail
  they fail because omitable tags may nest so more care is needed

* modified:   src/HTMLParser.js
* modified:   test/test-slowparse.js

pollei commented:
fixed dd and dt as per https://bugzilla.mozilla.org/show_bug.cgi?id=946393

``` html
<dl><dt>Coffee<dd>Black hot drink<dt>Milk<dd>White cold drink</dl>
```

got some ruby working

``` html
<ruby>漢<rp>(<rt>Kan<rp>)</rp>字<rp>(<rt>ji<rp>)</ruby>
```

added more tests of omitable tags; 6 of which currently fail
  they fail because omitable tags may nest so more care is needed

``` html
<select><optgroup label="X"><option>V<option>S<optgroup label="Z"><option>M<option>A</select>
<table><thead><tr><td><tbody><tr><td><tr><td><tfoot></table>
<ruby><rb>10<rb>31<rb>2002<rtc><rt>Month<rt>Day<rt>Year<rtc><rt>Expiration Date</ruby>
```

For example when optgroup Z is started option S and optgroup X both need to end.
In the table example I show that this nesting can be 3 levels easily.

 I'll need more work to make them pass. I've been reading through to see where I might need to add some loops and extra tests etc. It might be next weekend before I can try again.

I am happy Thimble can use slowparse again; I might slowly add things to it. My interest is mostly in  https://github.com/Khan/live-editor as I use their website a bit https://www.khanacademy.org/profile/spollei/ .
- modified:   src/HTMLParser.js
- modified:   test/test-slowparse.js

Pomax commented:
hm, in that case it's probably worth landing all of these separately, so that things that already work can be merged in, and things that require code changes so the tests pass can be worked on without holding up the other elements.

pollei committed:
Fixed a few problems with nested omitable tags; still has serious problems
down to 2 failures instead of 6
	modified:   src/DOMBuilder.js
	modified:   src/HTMLParser.js

pollei commented:
OK I was able to work on it again this weekend, but not finish it. down to only two failures. 
Most likely _popOmited is badly broken ... Hopefully next weekend, I'll have chance to debug things, and clean up some cruft.

Pomax commented on src/DOMBuilder.js:
```diff
@@ -35,6 +35,12 @@ module.exports = (function(){
       this.currentNode = this.currentNode.parentNode;
     },
+    popElements: function(n) {
+      while (n>0) {
```
since we're terminating on n=0, we can say `while(n--)` here, so that we don't need the decrement in the body.

Pomax commented:
So awsome! Looking forward to next weekend's result =D

flukeout commented:
Hey @pollei - it's only been two years, but guess what, we're adding Slowparse back to Thimble! You can check out the tracking issue here: https://github.com/mozilla/brackets/pull/876

Any chance that you are still going to finish this PR? Thanks!
